pkg/monitor/cluster/certificateexpirationstatuses_test.go

```diff
@@ -0,0 +1,141 @@
+package cluster
+
+import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"testing"
+	"time"
+
+	"github.com/golang/mock/gomock"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/kubernetes/fake"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+	mock_metrics "github.com/Azure/ARO-RP/pkg/util/mocks/metrics"
+	utiltls "github.com/Azure/ARO-RP/pkg/util/tls"
+)
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+func TestEmitCertificateExpirationStatuses(t *testing.T) {
+	expiration := time.Now().Add(time.Hour * 24 * 5)
+	expirationString := expiration.UTC().Format(time.RFC3339)
+	for _, tt := range []struct {
+		name            string
+		isManaged       bool
+		wantExpirations []map[string]string
+	}{
+		{
+			name:      "unmanaged domain",
+			isManaged: false,
+			wantExpirations: []map[string]string{
+				{
+					"subject":        "geneva.certificate",
+					"expirationDate": expirationString,
+				},
+			},
+		},
+		{
+			name:      "managed domain",
+			isManaged: true,
+			wantExpirations: []map[string]string{
+				{
+					"subject":        "geneva.certificate",
+					"expirationDate": expirationString,
+				},
+				{
+					"subject":        "contoso.aroapp.io",
+					"expirationDate": expirationString,
+				},
+				{
+					"subject":        "api.contoso.aroapp.io",
+					"expirationDate": expirationString,
+				},
+			},
+		},
+	} {
+		t.Run(tt.name, func(t *testing.T) {
+			ctx := context.Background()
+
+			var secrets []runtime.Object
+			_, genevaCert, err := utiltls.GenerateTestKeyAndCertificate("geneva.certificate", nil, nil, false, false, func(template *x509.Certificate) {
+				template.NotAfter = expiration
+			})
+			if err != nil {
+				t.Fatal(err)
+			}
+			secrets = append(secrets, &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "cluster",
+					Namespace: "openshift-azure-operator",
+				},
+				Data: map[string][]byte{
+					"gcscert.pem": pem.EncodeToMemory(&pem.Block{
+						Type:  "CERTIFICATE",
+						Bytes: genevaCert[0].Raw,
+					}),
+				},
+			})
+
+			var domain string
+			if tt.isManaged {
+				domain = "contoso.aroapp.io"
+				for _, sec := range []struct{ secretName, certSubject string }{
+					{
+						"foo12-ingress",
+						domain,
+					},
+					{
+						"foo12-apiserver",
+						"api." + domain,
+					},
+				} {
+					_, cert, _ := utiltls.GenerateTestKeyAndCertificate(sec.certSubject, nil, nil, false, false, func(template *x509.Certificate) {
+						template.NotAfter = expiration
+					})
+					secrets = append(secrets, &corev1.Secret{
+						ObjectMeta: metav1.ObjectMeta{
+							Name:      sec.secretName,
+							Namespace: "openshift-azure-operator",
+						},
+						Data: map[string][]byte{
+							"tls.crt": pem.EncodeToMemory(&pem.Block{
+								Type:  "CERTIFICATE",
+								Bytes: cert[0].Raw,
+							}),
+						},
+					})
+				}
```

Comment:
IMHO this is really difficult to follow as it defines an anonymous struct, creates a for over an instantiate slice with that struct + calls a function that defines an inline anonymous function. 
We can split it and reuse the previously defined function.

We can replace the above block code with:
```go
certsInfo := []certInfo{
	{
		"foo12-ingress",
		domain,
	},
	{
		"foo12-apiserver",
		"api." + domain,
	},
}
secretsFromCertsInfo := getSecretsFromCertsInfo(certsInfo, tweakTemplateFn)
secrets = append(secrets, secretsFromCertsInfo...)
```
defining also...
```go
type certInfo struct {
	secretName, certSubject string
}

func getSecretsFromCertsInfo(certsInfo []certInfo, tweakTemplateFn func(*x509.Certificate)) []runtime.Object {
	var secrets []runtime.Object
	for _, sec := range certsInfo {
		_, cert, _ := utiltls.GenerateTestKeyAndCertificate(sec.certSubject, nil, nil, false, false, tweakTemplateFn)
		s := &corev1.Secret{
			ObjectMeta: metav1.ObjectMeta{
				Name:      sec.secretName,
				Namespace: "openshift-azure-operator",
			},
			Data: map[string][]byte{
				"tls.crt": pem.EncodeToMemory(&pem.Block{
					Type:  "CERTIFICATE",
					Bytes: cert[0].Raw,
				}),
			},
		}
		secrets = append(secrets, s)
	}
	return secrets
}
```
I think after those changes the test is easier to follow 😃